File: middleware/a1/client/ProLight.py
from Ice import *

import devices_ice
from Devices import *


class ProLight:
    def __init__(self, config, communicator: CommunicatorI):
        self.name = config.split('=')[0]
        pro_light_proxy = communicator.propertyToProxy(self.name)
        self.proxy = IProLightPrx.checkedCast(pro_light_proxy)
        self.commands = ["getDeviceInfo", "toggleLightColor", "toggleLightIntensity", "setLightMode"]
        self.intensities = {"LOW":LightIntensity.LOW, "MEDIUM":LightIntensity.MEDIUM, "HIGH":LightIntensity.HIGH}
        self.colors = {"YELLOW":LightColor.YELLOW, "RED":LightColor.RED, "BLUE":LightColor.BLUE, "GREEN":LightColor.GREEN}
        if not self.proxy:
            print("Object not found for ", self.name)
            exit(1)

    def commandHandler(self, command: str):
        if command == "getDeviceInfo":
            info = self.proxy.getDeviceInfo()
            print("Data for device ", self.name, ":")
            print(info)

        elif command == "toggleLightColor":
            print(self.proxy.toggleLightColor())

        elif command == "toggleLightIntensity":
            print(self.proxy.toggleLightIntensity())

        elif command == "setLightMode":
            colors = ""
            for color in self.colors:
                colors += color + " "
            print("Choose one color from ", colors, ": ", end="")
            input_color = input()

            intensities = ""
            for intensity in self.intensities:
                intensities += intensity + " "
            print("Choose one intensity from ", intensities, ": ", end="")
            input_intensity = input()

            try:
                print(self.proxy.setLightMode(self.colors[input_color], self.intensities[input_intensity]))
            except KeyError:
                print("Wrong values provided, try again")
            except UnknownLightColor:
                print("Wrong light color entered, try again")
            except UnknownLightIntensity:
                print("Wrong light intensity entered, try again")

File: middleware/a1/server/src/gcyganek/main/lights/ProLight.java
package gcyganek.main.lights;

import Devices.*;
import com.zeroc.Ice.Current;

import java.util.Arrays;

public class ProLight extends BasicLight implements IProLight {

    public ProLight(String name) {
        super(name);

        this.mode.intensity = LightIntensity.LOW;
        this.mode.color = LightColor.RED;
    }

    @Override
    public String setLightMode(LightColor lightColor, LightIntensity lightIntensity, Current current) throws UnknownLightColor, UnknownLightIntensity {
        if (Arrays.stream(LightColor.values()).noneMatch(l -> l == lightColor)) {
            String reason = lightColor + " light color is not supported";
            throw new UnknownLightColor(reason);
        }

        if (Arrays.stream(LightIntensity.values()).noneMatch(i -> i == lightIntensity)) {
            String reason = lightIntensity + " light intensity is not supported";
            throw new UnknownLightIntensity(reason);
        }

        this.mode.intensity = lightIntensity;
        this.mode.color = lightColor;

        return "Light color set to " + mode.color + " and intensity set to " + mode.intensity;
    }
}
